server/cluster/metadata/topology_manager.go

```diff
@@ -390,6 +410,30 @@ func (m *TopologyManagerImpl) GetShardNodes() GetShardNodesResult {
 	}
 }
 
+func (m *TopologyManagerImpl) DropShardNodes(ctx context.Context, shardNodes []storage.ShardNode) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	newShardNodes := make([]storage.ShardNode, 0, len(m.clusterView.ShardNodes))
+
+	for _, originShardNode := range m.clusterView.ShardNodes {
+		if !contains(shardNodes, originShardNode) {
+			newShardNodes = append(newShardNodes, originShardNode)
+		}
+	}
+
+	return m.updateClusterViewWithLock(ctx, m.clusterView.State, newShardNodes)
+}
+
+func contains(shardNodes []storage.ShardNode, originShardNode storage.ShardNode) bool {
+	for _, dropShardNode := range shardNodes {
+		if originShardNode.NodeName == dropShardNode.NodeName && originShardNode.ID == dropShardNode.ID {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
When you use slice of a struct rather than a pointer, you have to be careful to not introduce copy when looping over it.

Reply:
Yes... maybe it should be written according to the principle similar to function visibility(Use private first in any case, unless the requirement cannot be met). In any case, use for index to traverse the slice first, unless it is found to avoid using a copy.
In the future, I will check all the slice traversal codes in the project individually, and fix them through a pull request.